File: app/src/main/java/pl/szczodrzynski/edziennik/ui/modules/login/LoginActivity.java
package pl.szczodrzynski.edziennik.ui.modules.login;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.navigation.NavDestination;
import androidx.navigation.NavOptions;
import androidx.navigation.Navigation;

import com.afollestad.materialdialogs.MaterialDialog;

import java.util.ArrayList;
import java.util.List;

import pl.szczodrzynski.edziennik.App;
import pl.szczodrzynski.edziennik.R;
import pl.szczodrzynski.edziennik.api.v2.models.ApiError;
import pl.szczodrzynski.edziennik.databinding.ActivityLoginBinding;
import pl.szczodrzynski.edziennik.ui.dialogs.error.ErrorSnackbar;

public class LoginActivity extends AppCompatActivity {

    private ActivityLoginBinding b;
    private App app;
    private static final String TAG = "LoginActivity";
    public static final int RESULT_OK = 1;
    public static NavOptions navOptions;

    static ApiError error = null;
    ErrorSnackbar errorSnackbar = new ErrorSnackbar(this);

    static List<LoginProfileObject> profileObjects;
    public static boolean firstCompleted = false; // if a profile is already added during *this* login. This means that LoginChooser has to navigateUp onBackPressed. Else, finish the activity.
    public static boolean privacyPolicyAccepted = false;

    @Override
    public void onBackPressed() {
        NavDestination destination = Navigation.findNavController(this, R.id.nav_host_fragment).getCurrentDestination();
        if (destination != null && destination.getId() == R.id.loginSyncErrorFragment) {
            return;
        }
        if (destination != null && destination.getId() == R.id.loginProgressFragment) {
            return;
        }
        if (destination != null && destination.getId() == R.id.loginSyncFragment) {
            return;
        }
        if (destination != null && destination.getId() == R.id.loginChooserFragment && !firstCompleted) {
            setResult(RESULT_CANCELED);
            finish();
            return;
        }
        if (destination != null && destination.getId() == R.id.loginSummaryFragment) {
            new MaterialDialog.Builder(this)
                    .title(R.string.are_you_sure)
                    .content(R.string.login_cancel_confirmation)
                    .positiveText(R.string.yes)
                    .negativeText(R.string.no)
                    .onPositive((dialog, which) -> {
                        setResult(RESULT_CANCELED);
                        finish();
                    })
                    .show();
            return;
        }
        Navigation.findNavController(this, R.id.nav_host_fragment).navigateUp();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setTheme(R.style.AppTheme_Light);

        firstCompleted = false;
        profileObjects = new ArrayList<>();
        error = null;

        navOptions = new NavOptions.Builder()
                .setEnterAnim(R.anim.slide_in_right)
                .setExitAnim(R.anim.slide_out_left)
                .setPopEnterAnim(R.anim.slide_in_left)
                .setPopExitAnim(R.anim.slide_out_right)
                .build();

        b = DataBindingUtil.inflate(getLayoutInflater(), R.layout.activity_login, null, false);
        setContentView(b.getRoot());

        errorSnackbar.setCoordinator(b.coordinator, null);

        app = (App) getApplication();

        if (!app.appConfig.loginFinished) {
            app.appConfig.miniDrawerVisible = getResources().getConfiguration().smallestScreenWidthDp > 480;
            app.saveConfig("miniDrawerVisible");
        }

        /*b.getRoot().addOnLayoutChangeListener(((v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
            Animator circularReveal = null;
            float finalRadius = (float) (Math.max(b.revealView.getWidth(), b.revealView.getHeight()) * 1.1);
            circularReveal = ViewAnimationUtils.createCircularReveal(b.revealView, b.revealView.getWidth()/2, b.revealView.getHeight()/2, 0, finalRadius);
            circularReveal.setDuration(400);
            circularReveal.setInterpolator(new AccelerateInterpolator());
            // make the view visible and start the animation
            b.revealView.setVisibility(View.VISIBLE);
            circularReveal.addListener(new Animator.AnimatorListener() {
                @Override public void onAnimationEnd(Animator animation) {
                    Anim.fadeIn(b.title1, 500, new Animation.AnimationListener() {
                        @Override public void onAnimationEnd(Animation animation) {
                            b.title2.postDelayed(() -> {
                                Anim.fadeIn(b.title2, 500, new Animation.AnimationListener() {
                                    @Override public void onAnimationEnd(Animation animation) {
                                        b.revealView.postDelayed(() -> {
                                            Anim.fadeOut(b.title1, 500, null);
                                            Animation anim2 = null;
                                            anim2 = new Anim.ResizeAnimation(b.revealView, 1.0f, 1.0f, 1.0f, 0.15f);
                                            anim2.setDuration(800);
                                            anim2.setInterpolator(new AccelerateDecelerateInterpolator());
                                            b.revealView.startAnimation(anim2);
                                        }, 700);
                                    }
                                    @Override public void onAnimationStart(Animation animation) { }
                                    @Override public void onAnimationRepeat(Animation animation) { }
                                });
                            }, 1500);
                        }
                        @Override public void onAnimationStart(Animation animation) { }
                        @Override public void onAnimationRepeat(Animation animation) { }
                    });
                }
                @Override public void onAnimationCancel(Animator animation) { }
                @Override public void onAnimationStart(Animator animation) { }
                @Override public void onAnimationRepeat(Animator animation) { }
            });
            circularReveal.start();
        }));*/

        /**/

        /*TextInputEditText e = findViewById(R.id.buttontest);

        e.setOnClickListener((v -> {
            Toast.makeText(this, "clicked", Toast.LENGTH_SHORT).show();
            PopupMenu popup = new PopupMenu(this, e);
            //popup.getMenu().add(0, 15, 0, HomeFragment.plural(c, R.plurals.time_till_seconds, 15));
            //popup.getMenu().add(0, 15 * 60, 0, HomeFragment.plural(c, R.plurals.time_till_minutes, 15));
            popup.getMenu().add(0, 30 * 60, 0, HomeFragment.plural(this, R.plurals.time_till_minutes, 30));
            popup.getMenu().add(0, 60 * 60, 0, HomeFragment.plural(this, R.plurals.time_till_hours, 1));
            popup.getMenu().add(0, 120 * 60, 0, HomeFragment.plural(this, R.plurals.time_till_hours, 2));
            popup.getMenu().add(0, 180 * 60, 0, HomeFragment.plural(this, R.plurals.time_till_hours, 3));
            popup.getMenu().add(0, 240 * 60, 0, HomeFragment.plural(this, R.plurals.time_till_hours, 4));
            popup.setOnMenuItemClickListener(item -> {
                e.setText(item.getTitle());
                return false;
            });
            popup.show();
        }));*/


    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/ui/login/LoginActivity.kt
/*
 * Copyright (c) Kuba Szczodrzyński 2020-4-16.
 */

package pl.szczodrzynski.edziennik.ui.login

import android.app.Activity
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.NavOptions
import androidx.navigation.Navigation
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import pl.szczodrzynski.edziennik.App
import pl.szczodrzynski.edziennik.R
import pl.szczodrzynski.edziennik.data.api.models.ApiError
import pl.szczodrzynski.edziennik.data.db.entity.LoginStore
import pl.szczodrzynski.edziennik.databinding.LoginActivityBinding
import pl.szczodrzynski.edziennik.ui.base.dialog.SimpleDialog
import pl.szczodrzynski.edziennik.ui.main.ErrorSnackbar
import kotlin.coroutines.CoroutineContext

class LoginActivity : AppCompatActivity(), CoroutineScope {
    companion object {
        private const val TAG = "LoginActivity"
    }

    private val app: App by lazy { applicationContext as App }
    private lateinit var b: LoginActivityBinding
    lateinit var navOptions: NavOptions
    lateinit var navOptionsBuilder: NavOptions.Builder
    val nav by lazy { Navigation.findNavController(this, R.id.nav_host_fragment) }
    val errorSnackbar: ErrorSnackbar by lazy { ErrorSnackbar(this) }
    val swipeRefreshLayout: SwipeRefreshLayout by lazy { b.swipeRefreshLayout }

    private val job: Job = Job()
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    var lastError: ApiError? = null
    val profiles = mutableListOf<LoginSummaryAdapter.Item>()
    val loginStores = mutableListOf<LoginStore>()

    fun getRootView() = b.root

    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        // TODO fix deprecation
        val destination = nav.currentDestination ?: run {
            nav.navigateUp()
            return
        }
        if (destination.id == R.id.loginSyncErrorFragment)
            return
        if (destination.id == R.id.loginProgressFragment)
            return
        if (destination.id == R.id.loginSyncFragment)
            return
        if (destination.id == R.id.loginFinishFragment)
            return
        // eggs
        if (destination.id == R.id.loginPrizeFragment) {
            finish()
            return
        }
        if (destination.id == R.id.loginChooserFragment && loginStores.isEmpty()) {
            setResult(Activity.RESULT_CANCELED)
            finish()
            return
        }
        if (destination.id == R.id.loginSummaryFragment) {
            SimpleDialog<Unit>(this) {
                title(R.string.are_you_sure)
                message(R.string.login_cancel_confirmation)
                positive(R.string.yes) {
                    setResult(Activity.RESULT_CANCELED)
                    finish()
                }
                negative(R.string.no)
            }.show()
            return
        }
        nav.navigateUp()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTheme(R.style.AppTheme_M3)

        navOptionsBuilder = NavOptions.Builder()
                .setEnterAnim(R.anim.slide_in_right)
                .setExitAnim(R.anim.slide_out_left)
                .setPopEnterAnim(R.anim.slide_in_left)
                .setPopExitAnim(R.anim.slide_out_right)
        navOptions = navOptionsBuilder.build()

        b = LoginActivityBinding.inflate(layoutInflater)
        setContentView(b.root)
        errorSnackbar.setCoordinator(b.coordinator, b.snackbarAnchor)

        app.buildManager.validateBuild(this)

        launch {
            app.config.loginFinished = app.db.profileDao().count > 0
            if (!app.config.loginFinished) {
                app.config.ui.miniMenuVisible = resources.configuration.smallestScreenWidthDp > 480
            }
        }
    }

    fun error(error: ApiError) { errorSnackbar.addError(error).show(); lastError = error }
}
